twoliter/src/lock.rs

```diff
@@ -28,3 +38,234 @@ pub(crate) struct LockedImage {
     /// The resolved image uri of the dependency
     pub source: String,
 }
+
+/// The hash should not contain the source to allow for collision detection
+impl Hash for LockedImage {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+        self.version.hash(state);
+        self.vendor.hash(state);
+    }
+}
+
+impl LockedImage {
+    // This is utilized to build a hashable comparator from an Image to a LockedImage
+    // for resolution of Twoliter.lock
+    #[allow(dead_code)]
+    fn hash_image(image: &Image) -> Self {
+        Self {
+            name: image.name.clone(),
+            version: image.version.clone(),
+            vendor: image.vendor.clone(),
+            source: String::default(),
+        }
+    }
+}
+
+#[derive(Deserialize, Debug)]
+#[allow(dead_code)]
+struct ImageMetadata {
+    /// The name of the kit
+    pub name: String,
+    /// The version of the kit
+    pub version: Version,
+    /// The required sdk of the kit,
+    pub sdk: Image,
+    /// Any dependent kits
+    pub kit: Vec<Image>,
+}
+
+#[derive(Deserialize, Debug)]
+#[allow(dead_code)]
+struct ManifestListView {
+    manifests: Vec<ManifestView>,
+}
+
+#[derive(Deserialize, Debug)]
+#[allow(dead_code)]
+struct ManifestView {
+    digest: String,
+}
+
+#[allow(dead_code)]
+impl Lock {
+    pub(crate) async fn new(project: &Project) -> Result<Self> {
+        let lock_file_path = project.project_dir().join("Twoliter.lock");
+        if lock_file_path.exists() {
+            let lock_str = read_to_string(&lock_file_path)
+                .await
+                .context("failed to read lockfile")?;
+            let lock: Self =
+                toml::from_str(lock_str.as_str()).context("failed to deserialize lockfile")?;
+            // If the digests match we don't need to re-resolve the lockfile
+            if lock.digest == project.digest()? {
+                return Ok(lock);
+            }
+        }
+        let lock = Self::resolve(&project).await?;
+        let lock_str = toml::to_string_pretty(&lock).context("failed to serialize lock file")?;
+        write(&lock_file_path, lock_str)
+            .await
+            .context("failed to write lock file")?;
+        Ok(lock)
+    }
+
+    async fn resolve(project: &Project) -> Result<Self> {
+        let mut kit: HashSet<LockedImage> = HashSet::new();
+        let vendor_table = project.vendor();
+        let mut sdk = if let Some(sdk) = project.sdk_image().as_ref() {
+            let sdk_vendor = vendor_table.get(&sdk.vendor).context(format!(
+                "vendor '{}' was not specified in Twoliter.toml, and is required for build",
+                sdk.vendor
+            ))?;
+            Some(LockedImage {
+                name: sdk.name.clone(),
+                version: sdk.version.clone(),
+                vendor: sdk.vendor.clone(),
+                source: format!("{}:{}/v{}", sdk_vendor.registry, sdk.name, sdk.version),
+            })
```

Comment:
The algorithm here could be cleaned up by unifying the treatment of kits and SDKs:
```
create a set of known vendors
create an empty map of known (name, vendor) -> version
create an empty vec of locked images

create a vec of images to resolve from the list of kits
if there's an SDK:
    add the SDK to the vec of images to resolve

while the vec of images to resolve is not empty:
    copy the vec of images to resolve to the working vec
    for image in working vec:
        if an entry for this image already exists in the version map:
            assert that the version is the same
            continue with the next image

        assert that the vendor is in the set of known vendors
        add the (name, vendor) -> version entry to the version map
        add the LockedImage to the vec of locked images

        add the image's sdk to the vec of images to resolve
        for each image dependency:
            add the dependency to the vec of images to resolve

finally:
    # error: must have an SDK
    assert that there's at least one SDK in the vec of locked images

    # error: cannot have more than one SDK
    assert that there's exactly one SDK in the vec of locked images
```

You are mostly doing this already, but you are not checking for cases where the same kit name/vendor is added at different versions, and the vendor lookup can be optimized.